src/desdemona/query.clj

```diff
@@ -0,0 +1,25 @@
+(ns desdemona.query
+  (:require [clojure.core.logic :as l]))
+
+(defn ^:private gen-query
+  "Expands a query and events to a core.logic program that executes
+  it."
+  [n-answers query events]
+  `(l/run ~n-answers [results#]
+     (l/fresh [~'x]
+       (l/== [~'x] results#)
+       (l/membero ~'x ~events)
+       ~query)))
+
+(defn run-query
+  "Runs a query over some events and finds n answers (default 1)."
+  ([query events]
+   (run-query 1 query events))
+  ([n-answers query events]
+   (let [compiled-query (gen-query n-answers query events)
+         old-ns *ns*]
+     (try
+       (in-ns 'desdemona.query)
+       (eval compiled-query)
+       (finally
+         (in-ns (ns-name old-ns)))))))
```

Comment:
While the `ns` macro is generally preferred over `in-ns`, this needs to be `in-ns` because the macro takes a symbol literal to determine the ns, and definitiionally we don't know that symbol. Alternatively, this could be done with `(binding [*ns* xyzzy] ...)`, but that wasn't done because that doesn't correctly make nses if they haven't been made yet and is generally... _more_ frowned upon than `in-ns` :)